About Dataset(From Kaggle)

"Superstore Sales" generally refers to the sales data and performance of a superstore or a large retail store that offers a wide range of products and merchandise. These superstores are typically known for their extensive inventory, competitive pricing, and one-stop shopping experience.

Analyzing superstore sales data is essential for store owners, managers, and analysts to understand the store's performance, identify trends, and make informed business decisions. Here are some key aspects of a US based superstore sales analysis.

- Row ID => Unique ID for each row.
- Order ID => Unique Order ID for each Customer.
- Order Date => Order Date of the product.
- Ship Date => Shipping Date of the Product.
- Ship Mode=> Shipping Mode specified by the Customer.
- Customer ID => Unique ID to identify each Customer.
- Customer Name => Name of the Customer.
- Segment => The segment where the Customer belongs.
- Country => Country of residence of the Customer.
- City => City of residence of of the Customer.
- State => State of residence of the Customer.
- Postal Code => Postal Code of every Customer.
- Region => Region where the Customer belong.
- Product ID => Unique ID of the Product.
- Category => Category of the product ordered.
- Sub-Category => Sub-Category of the product ordered.
- Product Name => Name of the Product
- Sales => Sales of the Product.
- Quantity => Quantity of the Product.
- Discount => Discount provided.
- Profit => Profit/Loss incurred.



In [ ]:
# importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# loading the dataset
df=pd.read_csv('/content/Sample - Superstore.csv', encoding='cp1252')
df

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9989,9990,CA-2014-110422,1/21/2014,1/23/2014,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,...,33180,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.2480,3,0.20,4.1028
9990,9991,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.9600,2,0.00,15.6332
9991,9992,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.5760,2,0.20,19.3932
9992,9993,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.6000,4,0.00,13.3200


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

In [ ]:
df.describe()

,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


In [ ]:
df.isnull().sum()

,0
Row ID,0
Order ID,0
Order Date,0
Ship Date,0
Ship Mode,0
Customer ID,0
Customer Name,0
Segment,0
Country,0
City,0


In [ ]:
df['Order Date'] = pd.to_datetime(df['Order Date'])

In [ ]:
df['Ship Date'] = pd.to_datetime(df['Ship Date'])

In [ ]:
print("Start Date:", df["Order Date"].min())
print("End Date:", df["Order Date"].max())

Start Date: 2014-01-03 00:00:00
End Date: 2017-12-30 00:00:00


In [ ]:
import sqlite3
conn = sqlite3.connect('superstore.db')
print("Database created successfully!")

Database created successfully!


In [ ]:
cursor=conn.cursor()

In [ ]:
df.to_sql(
    name='orders',
    con=conn,
    if_exists='replace',
    index=False
)

print("Data imported successfully!")

Data imported successfully!


In [ ]:
query = """
SELECT *
FROM orders
LIMIT 5;
"""

import pandas as pd

pd.read_sql(query, conn)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,2016-11-08 00:00:00,2016-11-11 00:00:00,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,2016-11-08 00:00:00,2016-11-11 00:00:00,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,2016-06-12 00:00:00,2016-06-16 00:00:00,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,2015-10-11 00:00:00,2015-10-18 00:00:00,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,2015-10-11 00:00:00,2015-10-18 00:00:00,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


# Business Problem 1: Overall Business Performance

**1. Total Sales Revenue**

In [ ]:
q1="""SELECT ROUND(SUM(Sales),2) AS Total_Sales
FROM orders;"""
df1 = pd.read_sql_query(q1,conn)
df1

,Total_Sales
0,2297200.86


Business Insight

The superstore generated $2297200.86 in total sales over the period 2014 to 2017 covered by the dataset.

**2. Total Profit**

In [ ]:
q2="""SELECT ROUND(SUM(Profit),2) AS Total_Profit
FROM orders;"""
df2=pd.read_sql_query(q2,conn)
df2

,Total_Profit
0,286397.02


Business Insight

The superstore earned a total profit of 286,397.02 during the analysis period. This indicates that while total sales were over $2.26 million, only about 12.66% of sales translated into profit.

**3. How many orders were placed?**

In [ ]:
q3="""SELECT COUNT(DISTINCT "Order ID") AS Total_Orders
FROM orders;"""
df3=pd.read_sql_query(q3,conn)
df3

,Total_Orders
0,5009


Business Insight

The superstore processed 5,009 unique customer orders

**4. How many unique customers purchased?**

In [ ]:
q4="""SELECT COUNT(DISTINCT "Customer ID") AS Total_Customers
FROM orders;"""
df4=pd.read_sql_query(q4,conn)
df4

,Total_Customers
0,793


Business Insight

A total of 793 unique customers purchased from the superstore during those 3 years.

**5. Average order value**

In [ ]:
q5="""SELECT
ROUND(SUM(Sales)/COUNT(DISTINCT "Order ID"),2) AS Avg_Order_Value
FROM orders;"""
df5=pd.read_sql_query(q5,conn)
df5

,Avg_Order_Value
0,458.61


Business Insight

The average order value (AOV) of $458.61 indicates the typical revenue generated per order. This metric helps evaluate customer purchasing behavior and can be used to measure the effectiveness of pricing, promotions, and cross-selling strategies.

# Business Problem 2: Product Performance

**6. Which product categories generate the highest sales?**

In [ ]:
q6="""SELECT
Category,
ROUND(SUM(Sales),2) AS Sales
FROM orders
GROUP BY Category
ORDER BY Sales DESC;"""
df6=pd.read_sql_query(q6,conn)
df6

,Category,Sales
0,Technology,836154.03
1,Furniture,741999.80
2,Office Supplies,719047.03


Business Insight

This indicates that Technology products were the primary driver of the Superstore's sales revenue during the analysis period.

**7. Which categories generate the highest profit?**

In [ ]:
q7="""SELECT
Category,
ROUND(SUM(Profit),2) AS Profit
FROM orders
GROUP BY Category
ORDER BY Profit DESC;"""
df7=pd.read_sql_query(q7,conn)
df7

,Category,Profit
0,Technology,145454.95
1,Office Supplies,122490.80
2,Furniture,18451.27


Business Insight

Technology has generated the highest profit being the category with highest sales but the major observation here is that even though the other cateogory furniture had higher sales than office supplies it generated a lesser profit than office supplies in the analysis period. This suggests that Furniture may have lower profit margins, higher discounts, or higher operational costs.

**8. Which sub-categories perform best?**

In [ ]:
q8="""SELECT
"Sub-Category",
ROUND(SUM(Sales),2) AS Sales
FROM orders
GROUP BY "Sub-Category"
ORDER BY Sales DESC;"""
df8=pd.read_sql_query(q8,conn)
df8

,Sub-Category,Sales
0,Phones,330007.05
1,Chairs,328449.10
2,Storage,223843.61
3,Tables,206965.53
4,Binders,203412.73
5,Machines,189238.63
6,Accessories,167380.32
7,Copiers,149528.03
8,Bookcases,114880.00
9,Appliances,107532.16


Business Insight

Phones was the highest-selling sub-category, generating 330,007.05 in sales, closely followed by Chairs with $328,449.10. These two sub-categories were the primary revenue drivers for the business.

Storage, Tables, and Binders also contributed significantly, each generating over $200,000 in sales.

On the other hand, Fasteners, Labels, and Envelopes recorded the lowest sales, indicating relatively low customer demand compared to other product groups.

**9. Which products generate the most revenue?**

In [ ]:
q9=""" SELECT "Product Name", ROUND(SUM(Sales),2) AS Sales
FROM Orders
GROUP BY "Product Name"
ORDER BY Sales DESC
LIMIT 10"""
df9=pd.read_sql_query(q9,conn)
df9

,Product Name,Sales
0,Canon imageCLASS 2200 Advanced Copier,61599.82
1,Fellowes PB500 Electric Punch Plastic Comb Bin...,27453.38
2,Cisco TelePresence System EX90 Videoconferenci...,22638.48
3,HON 5400 Series Task Chairs for Big and Tall,21870.58
4,GBC DocuBind TL300 Electric Binding System,19823.48
5,GBC Ibimaster 500 Manual ProClick Binding System,19024.50
6,Hewlett Packard LaserJet 3310 Copier,18839.69
7,HP Designjet T520 Inkjet Large Format Printer ...,18374.90
8,GBC DocuBind P400 Electric Binding System,17965.07
9,High Speed Automatic Electric Letter Opener,17030.31


Business Insight

The Canon imageCLASS 2200 Advanced Copier was the highest-selling product, generating $61,599.82 in sales, making it the store's top revenue contributor.

Office equipment and technology products dominate the top-selling products list, with copiers, printers, video conferencing systems, and binding machines contributing significantly to overall revenue.

The presence of multiple GBC, HP, Canon, and Cisco products among the top performers indicates strong customer demand for high-value office technology and business equipment.

**10. Which products are causing losses?**

In [ ]:
q10="""SELECT
"Product Name",
ROUND(SUM(Profit),2) AS Profit
FROM orders
GROUP BY "Product Name"
ORDER BY Profit
LIMIT 10;"""
df10=pd.read_sql_query(q10,conn)
df10

,Product Name,Profit
0,Cubify CubeX 3D Printer Double Head Print,-8879.97
1,Lexmark MX611dhe Monochrome Laser Printer,-4589.97
2,Cubify CubeX 3D Printer Triple Head Print,-3839.99
3,Chromcraft Bull-Nose Wood Oval Conference Tabl...,-2876.12
4,Bush Advantage Collection Racetrack Conference...,-1934.40
5,GBC DocuBind P400 Electric Binding System,-1878.17
6,Cisco TelePresence System EX90 Videoconferenci...,-1811.08
7,Martin Yale Chadless Opener Electric Letter Op...,-1299.18
8,Balt Solid Wood Round Tables,-1201.06
9,BoxOffice By Design Rectangular and Half-Moon ...,-1148.44


Business Insight

The Cubify CubeX 3D Printer Double Head Print recorded the highest loss, with a total profit of -$8,879.97, making it the least profitable product in the dataset.

Several high-value technology products and office furniture appear among the top loss-making products, indicating that expensive products do not always generate positive profits.

These losses may be attributed to heavy discounts, high procurement costs, low profit margins, or frequent returns. Further analysis is needed to determine the exact cause.

# Business Problem 3: Customer Analysis

**11. Top 10 customers**

In [ ]:
q11="""SELECT
"Customer Name",
ROUND(SUM(Sales),2) AS Sales
FROM orders
GROUP BY "Customer Name"
ORDER BY Sales DESC
LIMIT 10;"""
df11=pd.read_sql_query(q11,conn)
df11

,Customer Name,Sales
0,Sean Miller,25043.05
1,Tamara Chand,19052.22
2,Raymond Buch,15117.34
3,Tom Ashbrook,14595.62
4,Adrian Barton,14473.57
5,Ken Lonsdale,14175.23
6,Sanjit Chand,14142.33
7,Hunter Lopez,12873.30
8,Sanjit Engle,12209.44
9,Christopher Conant,12129.07


**12. Which customer segment contributes the highest revenue?**

In [ ]:
q12="""SELECT
Segment,
ROUND(SUM(Sales),2) AS Sales
FROM orders
GROUP BY Segment
ORDER BY Sales DESC;"""
df12=pd.read_sql_query(q12,conn)
df12

,Segment,Sales
0,Consumer,1161401.34
1,Corporate,706146.37
2,Home Office,429653.15


Business Insight

- Consumer is the largest customer segment by sales.
- Corporate customers are the second-largest source of revenue.
- Home Office contributes the smallest share of sales, indicating potential opportunities for targeted marketing and business growth.

**13. Which segment is most profitable?**

In [ ]:
q13="""SELECT Segment,ROUND(SUM(PROFIT),2) AS Profit
FROM orders
GROUP BY Segment
ORDER BY Profit DESC;"""
df13=pd.read_sql_query(q13,conn)
df13

,Segment,Profit
0,Consumer,134119.21
1,Corporate,91979.13
2,Home Office,60298.68


# Business Problem 4: Regional Performance

**14. Which region has the highest sales?**

In [ ]:
q14="""SELECT Region,ROUND(SUM(Sales),2) AS Sales
FROM orders
GROUP BY Region
ORDER BY Sales DESC;"""
df14=pd.read_sql_query(q14,conn)
df14

,Region,Sales
0,West,725457.82
1,East,678781.24
2,Central,501239.89
3,South,391721.91


Business Insight

- The West leads clearly, contributing about ~30% of total sales.
- East is a close second, suggesting strong demand in both coastal regions.
- Central and South lag behind, with South generating roughly half of West’s sales.

**15. Which state generates the highest revenue?**

In [ ]:
q15="""SELECT
State,
ROUND(SUM(Sales),2) AS Sales
FROM orders
GROUP BY State
ORDER BY Sales DESC
LIMIT 10;"""
df15=pd.read_sql_query(q15,conn)
df15

,State,Sales
0,California,457687.63
1,New York,310876.27
2,Texas,170188.05
3,Washington,138641.27
4,Pennsylvania,116511.91
5,Florida,89473.71
6,Illinois,80166.10
7,Ohio,78258.14
8,Michigan,76269.61
9,Virginia,70636.72


Business Insight

- California is the highest-performing state, making it the primary revenue driver.
- New York is the second-largest contributor, indicating strong sales in major markets.
- There is a significant drop in sales after the top two states, showing revenue is concentrated in a few locations.
- Texas and Washington are solid mid-performing states with potential for further growth.
- Others contribute relatively lower sales, presenting opportunities to expand market penetration and increase revenue.

# Business Problem 5: Sales Trends

**16. How do monthly sales change over time?**

In [ ]:
q16="""SELECT
strftime('%Y-%m',"Order Date") AS Month,
ROUND(SUM(Sales),2) AS Sales
FROM orders
GROUP BY Month
ORDER BY Month;"""
df16=pd.read_sql_query(q16,conn)
df16

,Month,Sales
0,2014-01,14236.90
1,2014-02,4519.89
2,2014-03,55691.01
3,2014-04,28295.35
4,2014-05,23648.29
5,2014-06,34595.13
6,2014-07,33946.39
7,2014-08,27909.47
8,2014-09,81777.35
9,2014-10,31453.39


# Business Problem 6: Discounts

**17. Does a higher discount reduce profit?**

In [ ]:
q17="""SELECT
Discount,
ROUND(AVG(Profit),2) AS Avg_Profit
FROM orders
GROUP BY Discount
ORDER BY Discount;"""
df17=pd.read_sql_query(q17,conn)
df17

,Discount,Avg_Profit
0,0.00,66.90
1,0.10,96.06
2,0.15,27.29
3,0.20,24.70
4,0.30,-45.68
5,0.32,-88.56
6,0.40,-111.93
7,0.45,-226.65
8,0.50,-310.70
9,0.60,-43.08


Business Insight

Profit is optimized at low discounts (0–10%), and anything above ~20% starts destroying profitability.

# Business Problem 7: Shipping

**18. Which shipping mode is used most?**

In [ ]:
q18="""SELECT
"Ship Mode",
COUNT(*) AS Orders
FROM orders
GROUP BY "Ship Mode"
ORDER BY Orders DESC;"""
df18=pd.read_sql_query(q18,conn)
df18

,Ship Mode,Orders
0,Standard Class,5968
1,Second Class,1945
2,First Class,1538
3,Same Day,543


Business Insight

Customers strongly prefer cost-effective shipping (Standard Class) over speed-based options.

**19. Which shipping mode generates the highest sales?**

In [ ]:
q19="""SELECT
"Ship Mode",
ROUND(SUM(Sales),2) AS Sales
FROM orders
GROUP BY "Ship Mode"
ORDER BY Sales DESC;"""
df19=pd.read_sql_query(q19,conn)
df19

,Ship Mode,Sales
0,Standard Class,1358215.74
1,Second Class,459193.57
2,First Class,351428.42
3,Same Day,128363.13


# Advanced

**20. Rank categories by sales**

In [ ]:
q20="""SELECT
Category,
SUM(Sales) AS Sales,
RANK() OVER(ORDER BY SUM(Sales) DESC) AS Rank
FROM orders
GROUP BY Category;"""
df20=pd.read_sql_query(q20,conn)
df20

,Category,Sales,Rank
0,Technology,836154.0330,1
1,Furniture,741999.7953,2
2,Office Supplies,719047.0320,3


**21. Customers whose sales are above the average customer sales**

In [ ]:
q21="""SELECT
"Customer Name",
SUM(Sales) AS Sales
FROM orders
GROUP BY "Customer Name"
HAVING SUM(Sales) >
(
SELECT AVG(CustomerSales)
FROM
(
SELECT SUM(Sales) AS CustomerSales
FROM orders
GROUP BY "Customer Name"
)
);"""
df21=pd.read_sql_query(q21,conn)
df21

,Customer Name,Sales
0,Aaron Smayling,3050.692
1,Adam Bellavance,7755.620
2,Adam Hart,3250.337
3,Adam Shillingsburg,3255.310
4,Adrian Barton,14473.571
...,...,...
289,Victoria Wilson,6134.038
290,William Brown,6160.102
291,Yana Sorensen,6720.444
292,Yoseph Carroll,5454.350


**22. Classify products by profitability**

In [ ]:
q22="""SELECT
"Product Name",
SUM(Profit) AS Profit,
CASE
WHEN SUM(Profit) > 500 THEN 'High Profit'
WHEN SUM(Profit) > 100 THEN 'Medium Profit'
ELSE 'Low Profit'
END AS Profit_Category
FROM orders
GROUP BY "Product Name";"""
df22=pd.read_sql_query(q22,conn)
df22

,Product Name,Profit,Profit_Category
0,"""While you Were Out"" Message Book, One Form pe...",10.3880,Low Profit
1,"#10 Gummed Flap White Envelopes, 100/Box",16.7678,Low Profit
2,#10 Self-Seal White Envelopes,52.1230,Low Profit
3,"#10 White Business Envelopes,4 1/8 x 9 1/2",223.1408,Medium Profit
4,"#10- 4 1/8"" x 9 1/2"" Recycled Envelopes",115.2806,Medium Profit
...,...,...,...
1845,iKross Bluetooth Portable Keyboard + Cell Phon...,115.6440,Medium Profit
1846,iOttie HLCRIO102 Car Mount,-11.9940,Low Profit
1847,iOttie XL Car Mount,-50.3748,Low Profit
1848,invisibleSHIELD by ZAGG Smudge-Free Screen Pro...,171.2648,Medium Profit


In [ ]:
import pandas as pd

with pd.ExcelWriter("Superstore_sales.xlsx", engine="openpyxl") as writer:

    # ---------------- Business Problem 1 ----------------
    start_row = 0
    for df in [df1, df2, df3,df4,df5]:
        df.to_excel(writer,
                    sheet_name="Problem_1",
                    startrow=start_row,
                    index=False)
        start_row += len(df) + 3   # Leave 3 blank rows

    # ---------------- Business Problem 2 ----------------
    start_row = 0
    for df in [df6, df7,df8,df9,df10]:
        df.to_excel(writer,
                    sheet_name="Problem_2",
                    startrow=start_row,
                    index=False)
        start_row += len(df) + 3

    # ---------------- Business Problem 3 ----------------
    start_row = 0
    for df in [df11, df12, df13]:
        df.to_excel(writer,
                    sheet_name="Problem_3",
                    startrow=start_row,
                    index=False)
        start_row += len(df) + 3

    # ---------------- Business Problem 4 ----------------
    start_row = 0
    for df in [df14, df15]:
        df.to_excel(writer,
                    sheet_name="Problem_4",
                    startrow=start_row,
                    index=False)
        start_row += len(df) + 3

    # ---------------- Business Problem 5 ----------------
    start_row = 0
    for df in [df16]:
        df.to_excel(writer,
                    sheet_name="Problem_5",
                    startrow=start_row,
                    index=False)
        start_row += len(df) + 3

    # ---------------- Business Problem 6 ----------------
    start_row = 0
    for df in [df17]:
        df.to_excel(writer,
                    sheet_name="Problem_6",
                    startrow=start_row,
                    index=False)
        start_row += len(df) + 3

    # ---------------- Business Problem 7 ----------------
    start_row = 0
    for df in [df18,df19]:
        df.to_excel(writer,
                    sheet_name="Problem_7",
                    startrow=start_row,
                    index=False)
        start_row += len(df) + 3

    # ---------------- Business Problem 8 ----------------
    start_row = 0
    for df in [df20, df21, df22]:
        df.to_excel(writer,
                    sheet_name="Problem_8",
                    startrow=start_row,
                    index=False)
        start_row += len(df) + 3

print("Excel file created successfully!")

Excel file created successfully!
